File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        """
        [1,2,3]
        """

        result: List[List[int]] = []
        def dfs(idx: int):
            if idx == len(nums):
                result.append(nums.copy())
                return

            for i in range(idx, len(nums)):
                nums[i], nums[idx] = nums[idx],nums[i]
                dfs(idx+1)
                nums[i], nums[idx] = nums[idx],nums[i]

        dfs(0)
        return result

File: Data Structures & Algorithms/subsets-ii/submission-1.cpp
class Solution {
public:
    vector<vector<int>>result;
    void dfs(vector<int>& nums,int curr,int n,vector<int>path){
        
        result.push_back(path);
        

        for(int i = curr; i<n; i++){
            if(i > curr && nums[i] == nums[i-1]) continue;

                path.push_back(nums[i]);

                dfs(nums,i+1,n,path);

                path.pop_back();
        }

    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        //all subsets backtracking and recursion
        result.clear();

        sort(nums.begin(),nums.end());
        int n = nums.size();
        dfs(nums,0,n,{});
        return result;
    }
};
